src/Voron/Impl/Journal/WriteAheadJournal.cs

```diff
@@ -576,6 +576,11 @@ public void OnTransactionCommitted(LowLevelTransaction tx)
             public long LastFlushedTransactionId => _lastFlushed.TransactionId;
             public long LastFlushedJournalId => _lastFlushed.JournalId;
             public long TotalWrittenButUnsyncedBytes => Interlocked.Read(ref _totalWrittenButUnsyncedBytes);
+
+            internal int TotalCommittedSinceLastFlushPages;
+            internal bool ShouldFlush => TotalCommittedSinceLastFlushPages != 0 || _lastFlushed.TransactionId != _waj._env.CurrentReadTransactionId; // TODO arek - double check the second condition
```

Comment:
Why the second part?

Reply:
The problem was described here - https://issues.hibernatingrhinos.com/issue/RavenDB-16536

TLDR we had this integer - `SizeOfUnflushedTransactionsInJournalFile` (which I renamed to `TotalCommittedSinceLastFlushPages`) which was used to track if we need to flush or not but it was decreased even before we actually started the flush operation. It didn't take into account that we might not be able to release all committed pages because there might be read transactions running during the flush.

I'm a bit afraid of tracking _exactly_ how many of modified pages has been flushed (there could be some overwrites of modified pages so we're filtering them out etc) so decided that the easiest would be to check if we still have something to flush by tracking the last flushed transaction ID. As commented need to carefully verify that.

Reply:
okay, understood